### ISSUE TITLE
py::isinstance<py::str>(py_bytes_object) should not return true

### ISSUE THREAD

**Issue by jbarlow83** [pybind/pybind11#1461]
## Issue description

In pybind 2.2.3, the template specialization `py::isinstance<py::str>()` returns True when used on `py::bytes` objects. This is incorrect in both Py2 and 3 (since `py::str` is `unicode` in Py2).

`py::instance(obj, py::str().get_type())` works correctly.

## Reproducible example code

```c++

    m.def("test_isinstance",
        []() {
            py::bytes b = "123";
            py::print(py::isinstance<py::str>(b));
        }
    );

```

```python

>>> pybind11.test_isinstance()
True

```

**Comment by jagerman** [comment]
The reason for the failure is that  `py::str`'s check function permissively allows both `str` and `bytes` objects (or under Python 2: both `unicode` and `str` objects).  This is mainly here for Python 2 compatibility to allow `str` objects to be automatically converted to `unicode` objects.

We can easily add a `isinstance<str>` specialization does only checks against `str` (Py 3) or `unicode` (Py 2) to keep the up-converting behaviour and fix `isinstance`.  That's going to work as expected under Python 3, but the behaviour under Python 2 is going to be a little strange:

```C++
m.def("f", [](py::object o) { return py::isinstance<py::str>(o); });
```
will result in:
```Python
m.f("abc")
```
returning `True` under Python 3 but `False` under Python 2, which might be a little surprising.

I don't think it's necessarily *wrong* for `isinstance` to work that way; just perhaps surprising.

**Comment by bstaletic** [comment]
Will be fixed in #2256
